Add Target and Ship class

alptuncs opened the pull request:
# Target Class
- [x] Target class oluşturma
- [x] Class fields
# Unit Tests
- [x] Target oluşturma
- [x] Ship oluşturma
- [x] Target Yerleştirme

alptuncs committed:
Add Board Class

Adding Board class and added unit tests for Board class functions

alptuncs committed:
Target class

Creating the target class and adding unit test

alptuncs committed:
Board Manager

Added a Board Manager class, renamed board to Board Renderer, added a render method to board renderer, added random ship place method to Board Manager, added unit tests for render and Board Manager

alptuncs committed:
Random Ship placement

modified the random ship placement so it also takes a target as parameter and randomly places it checking if there are any ships in the neighbour cells, added unit tests

alptuncs committed:
Adjacent Check

added adjacent check when randomly placing ships

SeferMirza commented on Battleships/Battleships/BoardManager.cs:
```diff
@@ -0,0 +1,240 @@
+
+        public void RandomPlaceShip(int count, Target ship)
+        {
+            if (count > 100)
+            {
+                throw new InvalidOperationException("Count can not exceed total cell count, total cell count = 100!");
+            }
```
Tek satırda yazarsan daha güzel gözükür

SeferMirza commented on Battleships/Battleships/Program.cs:
```diff
@@ -11,119 +11,29 @@ internal class Program
+            BoardRenderer boardRenderer = new BoardRenderer(10, 10);
+            BoardManager boardManager = new BoardManager();
+            boardRenderer.Render(boardManager);
+            Console.WriteLine("-----------------------------------------");
```
Console.WriteLine , ReadLine -> bu tarz metfotlar static methodlardır. Zorunda değilsen kullanmamak daha iyi olur yönetimide kolaylaştırır. Learn projesinde yaptığım şekilde `IConsole.cs` ve `SystemConsole.cs`  e bakabilirsin.

SeferMirza commented on Battleships/Battleships/Target.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    public class Target
+    {
+        private int xPos;
+        private int yPos;
+        private int size;
+        private string direction;
+
+
+
+        public int XPos => xPos;
+        public int YPos => yPos;
+        public int Size => size;
```
public int XPos { get; set; } şeklinde yaparsan bu yaptığınla aynı şey oluyor. VS de 'porp' yazıp tab tuşuna basarsan otomatik oluşturur.

SeferMirza commented on Battleships/Board Tests/BoardManagerUnitTests.cs:
```diff
@@ -0,0 +1,118 @@
+            {
+                Assert.IsTrue(e is InvalidOperationException);
+            }
+
```
Test içerisinde try catch yapmasan daha iyi bütün ortamı exception atacak şekilde ayarlayıp exception attığını test etmelisin.

alptuncs committed:
Board Manager

unit testler güncellendi, PlaceShip metodu eklendi,

alptuncs committed:
IConsole, Game manager

IConsole ve Game manager eklendi, exception unit testi düzeltildi

alptuncs committed:
Game manager

Game manager ve Board manager fonksiyonları eklendi

alptuncs committed:
unit tests

HasShip,IsHit unit testleri eklendi eski testleri adı değiştirildi

SeferMirza commented on Battleships/Board Tests/BoardRendererUnitTests.cs:
```diff
@@ -52,7 +52,7 @@ public void BoardTest3x2()
 
         [TestMethod]
-        public void PrintTest2x2()
+        public void Verilen_Boyutlarda_Oyun_Tahtası_Ekrana_Cizer()
```
Oyun tahtasını çizer, oyun tahtasını ekrana çizer diye test etmek yerine sadece tek test ile ekrana çizdirir şeklinde bir test case yazarsan zaten oyun tahtasının çizildiğini her türlü test etmiş olacaksın.

SeferMirza commented on Battleships/Board Tests/BoardRendererUnitTests.cs:
```diff
@@ -31,7 +31,7 @@ public void BoardTest2x3()
 
         [TestMethod]
-        public void BoardTest3x2()
+        public void Verilen_Boyutlarda_Oyun_Tahtası_Cizer_2()
```
1,2 diye gruplamak güzel durmuyor. Şundan büyük, şundan küçük olamaz diye test edip default bir değerlede çizdirsen kafi bence.
Örn: Testlerin üzerinde bir çizdirme methodu yap. Parametre olarakta boyutları al. Default olarakta 3,2 ver mesela. Ve çizdirme işlemini yaptığını tek test case ile doğrula. Örnek sadece sürece göre orta yolu sen bul tabi.

alptuncs replied:
testlerin üzerinde bir çizdirme methodu yapmayı tam anlamadım nasıl yapabilirim ?

SeferMirza replied:
```csharp
Board Draw(){
.....code
return board;
}
```

SeferMirza replied:
Testlerin üstünde olsa yeter. Learn içerisindeki testlerde AGame Amenu gibi örnekleri mevcut

SeferMirza commented on Battleships/Battleships/BoardRenderer.cs:
```diff
@@ -0,0 +1,79 @@
+                }
+                _boardGraphicString += i < BoardSurface.Length - 1 ? "\n" : "";
+            }
+            Console.WriteLine(_boardGraphicString);
```
Static olan Console sanırım bu. Bunu normal IConsole türündeki çevirelim yada return ile string döndürüp uygun yerde IConsole daki writeline ile bastıralım

alptuncs replied:
console ile değiştirdim

SeferMirza commented on Battleships/Board Tests/BoardManagerUnitTests.cs:
```diff
@@ -0,0 +1,86 @@
+        }
+        [TestMethod]
+
+        public void Tahta_Dısı_Koordinat_Verilince_IsHit_Exception_Atar()
```
türkçe karakter kullanmayalım

alptuncs replied:
düzelttim

alptuncs committed:
unit test

unit testler düzeltildi

SeferMirza commented on Battleships/Battleships/Program.cs:
```diff
@@ -38,8 +38,8 @@ static void Main(string[] args)
 
-            gameManager.FireMissile(playerBoard, 5, 5);
-            gameManager.FireMissile(computerBoard, 5, 5);
+            //gameManager.FireMissile(playerBoard, 5, 5);
+            //gameManager.FireMissile(computerBoard, 5, 5);
```
Yorum satırıyla kod bırakmayalım

SeferMirza commented on Battleships/Battleships/BoardRenderer.cs:
```diff
@@ -20,7 +21,7 @@ public BoardRenderer(int height, int width)
             this._height = height;
             this._width = width;
-
+            console = new SystemConsole();
```
Bu class içerisine consolu eklemeye gerek yok bence farklı bir yöntemle çözebilirsin. İçeriğe ters. Bu class ın amacına baktığımızda oyunun matematiği burada yapılıyor. Ekrana bir şey bastırma input alma gibi işlemler bu class ı ilgilendirmemeli.

SeferMirza commented:
Yukarıdaki tasklardan yaptığını düşündüğün taskı işaretlemeyi unutma

SeferMirza commented on Battleships/Board Tests/BoardRendererUnitTests.cs:
```diff
@@ -8,48 +8,7 @@ namespace Board_Tests
+       
 
         [TestMethod]
         public void Verilen_Boyutlarda_Oyun_Tahtası_Ekrana_Cizer()
```
Türkçe karakter mevcut

SeferMirza commented on Battleships/Board Tests/BoardRendererUnitTests.cs:
```diff
@@ -62,18 +21,6 @@ public class BoardRendererUnitTests
-        }
 
         [TestMethod]
         public void Verilen_Boyutlarda_Oyun_Tahtasını_Render_Fonksiyonu_ile_Ekrana_Cizer()
```
Türkçe karakter mevcut

alptuncs committed:
Unit tests, Iconsole

IConsole boardrenderer dan çıkarıldı, Boardrenderer bir şey çizmiyor string return ediyor, Game manager tahtayı ekrana çiziyor

alptuncs committed:
BoardRenderer

silmeyi unuttuğum 1 satırı sildim

SeferMirza commented on Battleships/Battleships/BoardManager.cs:
```diff
@@ -0,0 +1,306 @@
+            {
+                if (CheckPlacementBounds(x, y, ship))
+                {
+                    if (HasShip(x, y) || !CheckAdjacentSquares(x, y))
+                    {
+                        return false;
+                    }
```
Tek satıra alalım

SeferMirza replied:
Tek satır yapabildiğin yerler varsa bu tarz onlara bakıp yaparsan güzel olur ben sadece 1 tanesini gösterdim tüm projeyi kapsıyor.

SeferMirza commented on Battleships/Battleships/BoardManager.cs:
```diff
@@ -0,0 +1,306 @@
+                    return true;
+                }
+            }
+
```
if else arası boşluk olmasın bunlar tek yapılar

SeferMirza commented on Battleships/Battleships/BoardManager.cs:
```diff
@@ -0,0 +1,306 @@
+                        return false;
+                    }
+                }
+
```
Bu tarz boşlukları silelim okunmayı zorlaştırıyor bir çok insan tercih etmez

alptuncs committed:
Boşluk Silme

Class içinde olan gereksiz boşluklar silindi

alptuncs committed:
Unit Test Boşluk Silme
